ext/drreg/drreg.c

```diff
@@ -2047,43 +2063,78 @@ drreg_event_restore_state_with_ilist(void *drcontext, bool restore_memory,
      * different slot) after app instr write, multiple reg restores due to app
      * instr read, aflags spilled to reg instead of slot (for x86 only), lazy
      * restores, multi-phase use of drreg leading to nested or overlapping
```

Comment:
Starting an independent thread for this: How are multiple restore state events handled? If multiple clients and core DR had altered some machine state (e.g. spilled some gpr to some location that they own), then all of them will try to restore it in their respective restore state events. Does the order in which we invoke the restore state events guarantee that the gpr's final restored value will be the app value?

Reply:
DR always goes first and restores state that was changed in its mangling or instrumentation it owns like clean calls.  Clients/extensions are in order of the callback registration or drmgr priority.